tests/generators/generators.js

```diff
@@ -0,0 +1,107 @@
+function *stmt_yield(): Generator<number, void, void> {
+  yield 0; // ok
+  yield ""; // error: string ~> number
+}
+
+function *stmt_next(): Generator<void, void, number> {
+  (yield undefined : number); // ok
+  (yield undefined : string); // error: number ~> string
+}
+
+function *stmt_return_ok(): Generator<void, number, void> {
+  return 0; // ok
+}
+
+function *stmt_return_err(): Generator<void, number, void> {
+  return ""; // error: string ~> number
+}
+
+function *infer_stmt() {
+  var x: boolean = yield 0;
+  return "";
+}
+for (var x of infer_stmt()) { (x : string) } // error: number ~> string
+var infer_stmt_next = infer_stmt().next(0).value; // error: number ~> boolean
+if (typeof infer_stmt_next === "undefined") {
+} else if (typeof infer_stmt_next === "number") {
+} else {
+  (infer_stmt_next : boolean) // error: string ~> boolean
+}
+
+declare function *decl(): Generator<number, string, boolean>;
+for (var x of decl()) { (x : string) } // error: number ~> string
+var decl_next = decl().next(0).value; // error: number ~> boolean
+if (typeof decl_next === "undefined") {
+} else if (typeof decl_next === "number") {
+} else {
+  (decl_next : boolean) // error: string ~> boolean
+}
+
+function *widen_next() {
+  var x = yield 0;
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+widen_next().next(0)
+widen_next().next("")
+widen_next().next(true)
+
+function *widen_yield() {
+  yield 0;
+  yield "";
+  yield true;
+}
+for (var x of widen_yield()) {
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+
+function *delegate_next_generator() {
+  function *inner() {
+    var x: number = yield undefined; // error: string ~> number
+  }
+  yield *inner();
+}
+delegate_next_generator().next("");
+
+function *delegate_yield_generator() {
+  function *inner() {
+    yield "";
+  }
+
+  yield *inner();
+}
+for (var x of delegate_yield_generator()) {
+  (x : number) // error: string ~> number
+}
+
+function *delegate_return_generator() {
+  function *inner() {
+    return "";
+  }
+
+  var x: number = yield *inner() // TODO: error: string ~> number
+}
+
+// only generators can make use of a value passed to next
+function *delegate_next_iterable(xs: Array<number>) {
+  yield *xs;
+}
+delegate_next_iterable([]).next(""); // TODO: error: string ~> void
```

Comment:
If I understand correctly, you would like to raise an error when someone does something like

```JavaScript
function *foo() { yield *[1]; }
foo().next('bar');
```

even though this isn't a runtime error. If I'm reading the spec correctly, this will basically call 

```JavaScript
[1][Symbol.iterator]().next('bar')
```

and the next function [expects zero arguments](https://github.com/facebook/flow/blob/3f6411d58d6a666dba5fdc3bb4065cf72a44982c/lib/core.js#L303).

In Flow so far, it's not an error to call a function with more arguments than it expects, mainly because people do this in JS so bloody often. So you probably could just ignore next altogether if you're dealing with an `Iterable` and avoid the problem completely.

I think the error will never be as clear as you like if you're using an intersection type. If none of the intersected types match, it's really hard to figure out the right error message. We could explain that each type doesn't fit or we can explain that the whole intersection type doesn't fit.

I think what you want for a good error message (if you decide to ban passing things to next when you delegate to iterables) is a custom use type that behaves like an intersection, except that it makes it's decision exclusively based on `yield *`'s argument. You would flow the type of the expression to your custom use type, into which you'd pack all the info you might possibly need. When that lower bound finally resolves to something concrete, you would check if it's a `Generator` (and if so flow your next & return types around) or if it's an `Iterable` (and make sure next is void) or if it's something else (and blow up).

Now that I think about it, I don't think the error you are seeking is a good thing. I can construct an `Iterable` object who's `Iterator`'s `next()` method takes an argument. Here is an example:

```JavaScript
var MyIterable = {
  [Symbol.iterator]: function *() { return yield 4; }
}
```

I tested out the behavior in [Babel](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&playground=true&code=function%20*simpleGen(x)%20%7B%20return%20yield%20x%3B%20%7D%0A%0Avar%20myThing%20%3D%20%7B%0A%20%20%5BSymbol.iterator%5D%3A%20()%20%3D%3E%20simpleGen(1)%0A%7D%3B%0A%0Afunction%20*foo()%20%7B%0A%20%20console.log(%22delegate%20returned%22%2C%20yield%20*myThing)%3B%0A%7D%0A%0Avar%20g1%20%3D%20foo()%3B%0Aconsole.log(g1.next('firstNext'))%0A%0A%2F%2F%20Honestly%20I'm%20surprised%20this%20is%20what%20is%20returned%0Aconsole.log(g1.next('secondNext'))%0A%0A%0Afunction%20*bar()%20%7B%0A%20%20console.log(%22delegate%20returned%22%2C%20yield%20*simpleGen(2))%3B%0A%7D%0A%0Avar%20g2%20%3D%20bar()%3B%0Aconsole.log(g2.next('firstNext'))%0A%0A%2F%2F%20Again%2C%20I'm%20surprised%20this%20is%20what%20is%20returned%0Aconsole.log(g2.next('secondNext'))) and also Chrome so it is possible to pass a `next` through to an `Iterable`'s `Iterator`. This stuff is pretty complicated :P